### PR TITLE
[Merged by Bors] - feat: port Data.Fin.Tuple.NatAntidiagonal

### PR DESCRIPTION
**Pull request by mo271**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Fin/Tuple/NatAntidiagonal.lean [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+theorem mem_antidiagonalTuple {n : ℕ} {k : ℕ} {x : Fin k → ℕ} :
+    x ∈ antidiagonalTuple k n ↔ (∑ i, x i) = n := by
+  induction x using Fin.consInduction generalizing n with
+  | h0 =>
+    cases n
+    · simp
+    · simp [eq_comm]
+  | h x₀ x ih =>
+    simp_rw [Fin.sum_cons]
+    rw [antidiagonalTuple]  -- porting note: simp_rw doesn't use the equation lemma properly
```
Comment (eric-wieser) [inline Mathlib/Data/Fin/Tuple/NatAntidiagonal.lean:93]: This is worth raising on Zulip

File: Mathlib/Data/Fin/Tuple/NatAntidiagonal.lean [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+namespace List.Nat
+
+/-- `List.antidiagonalTuple k n` is a list of all `k`-tuples which sum to `n`.
+
+This list contains no duplicates (`List.Nat.nodup_antidiagonalTuple`), and is sorted
+lexicographically (`List.Nat.antidiagonalTuple_pairwise_pi_lex`), starting with `![0, ..., n]`
+and ending with `![n, ..., 0]`.
+
+```
+#eval antidiagonal_tuple 3 2
```
Comment (eric-wieser) [inline Mathlib/Data/Fin/Tuple/NatAntidiagonal.lean:62]:
```suggestion
#eval antidiagonalTuple 3 2
```

File: Mathlib/Data/Fin/Tuple/NatAntidiagonal.lean (hunk 1 above)
Comment (mo271) [reply Mathlib/Data/Fin/Tuple/NatAntidiagonal.lean:93]: https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/simp_rw.20not.20working.20as.20expected

File: Mathlib/Data/Fin/Tuple/NatAntidiagonal.lean [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+
+theorem antidiagonalTuple_two (n : ℕ) :
+    antidiagonalTuple 2 n = (antidiagonal n).map (piFinTwoEquiv fun _ => ℕ).symm.toEmbedding :=
+  Finset.eq_of_veq (Multiset.Nat.antidiagonalTuple_two n)
+#align finset.nat.antidiagonal_tuple_two Finset.Nat.antidiagonalTuple_two
+
+section EquivProd
+
+/-- The disjoint union of antidiagonal tuples `Σ n, antidiagonalTuple k n` is equivalent to the
+`k`-tuple `fin k → ℕ`. This is such an equivalence, obtained by mapping `(n, x)` to `x`.
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/Fin/Tuple/NatAntidiagonal.lean:271]:
```suggestion
`k`-tuple `Fin k → ℕ`. This is such an equivalence, obtained by mapping `(n, x)` to `x`.
```

**Review by Ruben-VandeVelde — APPROVED** [review]

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**